### ISSUE TITLE
Plugins write their settings to storage, even when nothing changed.

### ISSUE THREAD

**Issue by Gitoffthelawn** [notepad-plus-plus/notepad-plus-plus#9408]
### Description of the Issue
Plugins write their settings to storage, even when nothing changed.

### Steps to Reproduce the Issue

1. Get a tool to observe all file read/writes in real-time.  For example, this is excellent and free (but not currently open-source): https://www.nirsoft.net/utils/file_activity_watch.html
2. Run Notepad++
3. Exit Notepad++
4. Observe output from step 1's tool

### Expected Behavior
Plugins not to write anything to storage if nothing has changed.

### Actual Behavior
Many plugins write to storage even when nothing has changed.  One example is TextFX, which is included with Notepad++.  Another is the popular Hex Editor plugin.  I can provide a list of more plugins with this behavior, upon request.  The biggest question: Is this an issue with each of these plugins, a Notepad++ issue, or both?

### Debug Information
Notepad++ 7.9.2 32-bit
Windows 64-bit

**Comment by dail8859** [comment]
Notepad++ doesn't control or provide any features to plugins on how to save settings. So it is completely on the plugin side.

**Comment by Gitoffthelawn** [comment]
@dail8859 Thanks.  It's amazing (and surprising) how many Notepad++ plugins have this issue.

What to do about the TextFX plugin, since that is included with Notepad++?  It has this issue, and is included in Notepad++ by default.  Looking at its repo, it hasn't been updated for over a decade.

**Comment by dail8859** [comment]
> since that is included with Notepad++

As far as I know it has _not_ been included with Notepad++ by default for quite a while.

> It's amazing (and surprising) how many Notepad++ plugins have this issue.

My two cents from a plugin author's perspective...I've always found it far simpler to just flush out settings each time Notepad++ is closed instead of tracking when something may have changed.

**Comment by Gitoffthelawn** [comment]
> As far as I know it has _not_ been included with Notepad++ by default for quite a while.

I just checked, and you are correct!  I've been using Notepad++ for quite some time, and I didn't know TextFX was removed.  A bit OT, but do you know why it was removed, and if there is something similar, but better now?

> My two cents from a plugin author's perspective...I've always found it far simpler to just flush out settings each time Notepad++ is closed instead of tracking when something may have changed.

Your two cents are very insightful and helpful!  Yes, it is simpler, indeed.  But when the user has many plugins installed, it causes a whole bunch of non-sequential writing to non-volatile storage when Notepad++ closes.  The issue is amplified by Notepad++ doing this too.  Further amplifying the issue is that many security apps then need to process *all* those writes to help validate their safety.  Best practice is to not write if not needed, but as you mention, it involves a bit more coding.

**Comment by dail8859** [comment]
> A bit OT, but do you know why it was removed,

Not 100% sure but I believe due to bugs and lack of a maintainer. At one point it was not ported to 64 bit either. I think the project is in a bit better shape now but can't say for certain.

> Further amplifying the issue is that many security apps then need to process all those writes to help validate their safety.

In the grand scheme of your entire operating system and all other software writing files, logs, temp files etc, this would be a minuscule amount of IO that would need to be checked. I would hope if a particular plugin knows it has to do some heavy IO it would take those kinds of extra precautions.

So in the end it would be best to contact plugin maintainers to see if they would want to address these issues in their plugin.

**Comment by Gitoffthelawn** [comment]
@dail8859 wrote:
> Not 100% sure but I believe due to bugs and lack of a maintainer. At one point it was not ported to 64 bit either. I think the project is in a bit better shape now but can't say for certain.

I did a little research, and it looks like you are spot on.  Bugs, and no one maintaining.  Also, lack of 64-bit, as you mentioned.  If you think of anything comparable that is maintained, I'm interested.

> In the grand scheme of your entire operating system and all other software writing files, logs, temp files etc, this would be a minuscule amount of IO that would need to be checked. I would hope if a particular plugin knows it has to do some heavy IO it would take those kinds of extra precautions.

Yes, it's tiny in the grand scheme, for sure.  But when you use a bunch of Notepad++ plugins, and they are all trying to write settings at the same time when Notepad++ closes, it can get a bit laggy.  Sometimes I find myself opening an alternative notepad program, Metapad, simply because Meta[ad is much snappier to open and close as compared to Notepad++.  (Also Metapad has the benefit of displaying binary files in a more readable format... a bit closer to the Linux `strings` command.)

> So in the end it would be best to contact plugin maintainers to see if they would want to address these issues in their plugin.

That sounds like great advice.  Thanks.  I just added that task to my list of issues to report to various open-source projects.  Currently over 2000 items on that list... so it may be a while until I get time to write those specific issue reports!
